DOC-5371 - two phase finality documentation and release notes (#1655)
corda/corda-docs-portal: content/en/platform/corda/4.11/enterprise/release-notes-enterprise.md

## New features and enhancements

### Two Phase Finality

Two Phase Finality protocol (`FinalityFlow` and `ReceiveFinalityFlow` sub-flows) has been added to improve resiliency and

> recoverability of CorDapps using finality. Existing CorDapps will not require any changes to take advantage of this

Reviewer: ```suggestion
recoverability of CorDapps using finality. Existing CorDapps do not require any changes to take advantage of this
```

new improved protocol.

For more information, see [Two Phase Finality]({{< relref "two-phase-finality.md" >}})

### Improved double-spend exception handling

> Two Phase Finality will automatically delete an unnotarized transaction from the `DBTransaction` table if a double spend

Reviewer: ```suggestion
Two Phase Finality automatically deletes an unnotarized transaction from the `DBTransaction` table if a double spend
```

Author: Done

is detected upon attempting notarization by the the initiator of `FinalityFlow`.

Additionally, if the new optional `ReceiveFinalityFlow` `handlePropagatedNotaryError` constructor parameter is set to `true` (default: `false`),

> then the double spend error (NotaryError.Conflict) will propagate back to the 2PF initiator. This enables the initiator to automatically remove the associated unnotarized transaction from its `DBTransaction` table. 

Reviewer: ```suggestion
then the double spend error (NotaryError.Conflict) propagates back to the 2PF initiator. This enables the initiator to automatically remove the associated unnotarized transaction from its `DBTransaction` table. 
```

Author: Done

### Deserializing AMQP data performance improvement

This release includes improvements in the performance of deserializing AMQP data, which may result in performance improvements for LedgerGraph, Archiving and other CorDapps.

### Detect vault changes while vault query pages loaded

A new property, `previousPageAnchor`, has been added to `Vault.Page`. It is used to detect if the vault has changed while pages of a vault query have been loaded. If such a scenario is important to detect, then the property can be used to restart querying.

A example of how to use this property can be found in [Vault Queries]({{< relref "cordapps/api-vault-query.md#query-for-all-states-using-a-pagination-specification-and-iterate-using-the-totalstatesavailable-field-until-no-further-pages-available-1" >}}).

### Upgraded dependencies 
  
The following dependencies have been upgraded to address critical and high-severity security vulnerabilities:
* H2 has been upgraded from 1.4.197 to 2.1.214.
* Hibernate has been upgraded from 5.4.32.Final to 5.6.14.Final.
* Liquibase has been upgraded from 3.6.3 to 4.20.0.

### Consuming transaction IDs added to `vault_state` table 

> When a state is consumed by a transaction, Corda now adds the ID of the consuming transaction in the `consuming_tx_id` column of the `vault_state` table . Corda only updates this database column for new transactions; for existing consumed states already in the ledger, the value of `consuming_tx_id` is null.

Reviewer: ```suggestion
When a state is consumed by a transaction, Corda now adds the ID of the consuming transaction in the `consuming_tx_id` column of the `vault_state` table. Corda only updates this database column for new transactions; for existing consumed states already in the ledger, the value of `consuming_tx_id` is null.
```

Author: Done

### Two Phase Finality

Two Phase Finality protocol (`FinalityFlow` and `ReceiveFinalityFlow` sub-flows) has been added to improve resiliency and
recoverability of CorDapps using finality. Existing CorDapps do not require any changes to take advantage of this
new improved protocol.

For more information, see [Two Phase Finality]({{< relref "two-phase-finality.md" >}})

### Improved double-spend exception handling

Two Phase Finality automatically deletes an unnotarized transaction from the `DBTransaction` table if a double spend
is detected upon attempting notarization by the the initiator of `FinalityFlow`.

Additionally, if the new optional `ReceiveFinalityFlow` `handlePropagatedNotaryError` constructor parameter is set to `true` (default: `false`),

> then the double spend error (NotaryError.Conflict) propagates back to the 2PF initiator. This enables the initiator to automatically remove the associated unnotarized transaction from its `DBTransaction` table. 

Reviewer: ```suggestion
then the double spend error (`NotaryError.Conflict`) propagates back to the 2PF initiator. This enables the initiator to automatically remove the associated unnotarized transaction from its `DBTransaction` table. 
```
